File: matlab/common/TileAnalysis.m
%% header
% the tile analysis is to put all stocks into equally sized groups and the number of groups is defined by Tile
% and then construct tile portfolios for each tile either equally or cap-weighted.
% analyze the time series returns of each tile portfolio

% Date: 8/2/2015

%%

classdef TileAnalysis < handle
    
    properties
        NumTiles       % the number of tiles
        Dates          % the testing dates 
        UnivPP         % universe portfolio provier
    end
    
    methods
        function obj = TileAnalysis(dates, univPP, numTiles)
            if ischar(dates)
                obj.Dates = datenum(dates,'yyyymmdd');
            else
                obj.Dates = dates;
            end
            
            obj.UnivPP = univPP;
            
            if (nargin>2) 
                obj.NumTiles = numTiles;
            else
                obj.NumTiles = 5;
            end
        end
        
        
        function [topTileRets, botTileRets, spreads] = Run(obj, factor, bmDemean)
            periods = length(obj.Dates)-1;
            topTileRets = ReturnSeries();
            botTileRets = ReturnSeries();
            spreads = ReturnSeries();
            
            BMDemean = false;
            if(nargin>2 && bmDemean==true)
                BMDemean = true;
            end
            
            for i = 1:periods
                dt = obj.Dates(i);
                nextdt = obj.Dates(i+1);
                portfolio = GetPortfolioAsofFixed(obj.UnivPP, dt, 0);
                stockIDs = GetStockIDs(portfolio);
                %scores = arrayfun(@(stkid) GetScore(BP, stkid,dt),stockIDs);  doesn't work due to string array treated as char matrix
                
                % calculate factor scores and sort them
                totNumStks = length(stockIDs);
                scores = zeros(totNumStks,1);
                for j=1:totNumStks
                    scores(j) = GetScore(factor, stockIDs(j,:),dt, 1);
                end
                
                % nan scores are on the top
                [scoresSortedDown, idx] = sort(scores,'descend');
                numNaNs = sum(isnan(scores));
                if(numNaNs/totNumStks > 0.5)
                    error('Half of stocks have NaN scores')
                end
                
                numStkInTile = floor((totNumStks-numNaNs) / obj.NumTiles);
                
                topTileStkRets = zeros(numStkInTile,1);
                botTileStkRets = zeros(numStkInTile,1);
                for j=1:numStkInTile
                    topTileStkRets(j) = TotalReturnInRange_Bk(Stock.ByWindID(stockIDs(idx(j+numNaNs),:)), dt, nextdt);
                    botTileStkRets(j) = TotalReturnInRange_Bk(Stock.ByWindID(stockIDs(idx(totNumStks-j+1),:)), dt, nextdt);
                end
                
                % equal weighted for now, later need to check whether it is
                % cap weighted or equal weighted
                topRet = nanmean(topTileStkRets);
                botRet = nanmean(botTileStkRets);
                
                if(BMDemean)
                    bmRet = TotalReturnInRange_Bk(obj.UnivPP, dt, nextdt);
                    topRet = topRet - bmRet;
                    botRet = botRet - bmRet;
                end
                
                Add(topTileRets, nextdt, topRet);
                Add(botTileRets, nextdt, botRet);
                Add(spreads, nextdt, topRet - botRet);
            end
            
            disp(['Top AnnMean, ', num2str(topTileRets.AnnMean)]);
            disp(['Top AnnStd,  ', num2str(topTileRets.AnnStd)]);
            disp(['Top SR,      ', num2str(topTileRets.SR)]);
            
            disp(['Bot AnnMean, ', num2str(botTileRets.AnnMean)]);
            disp(['Bot AnnStd,  ', num2str(botTileRets.AnnStd)]);
            disp(['Bot SR,      ', num2str(botTileRets.SR)]);
            
            disp(['Spread AnnMean, ', num2str(spreads.AnnMean)]);
            disp(['Spread AnnStd,  ', num2str(spreads.AnnStd)]);
            disp(['Spread SR,      ', num2str(spreads.SR)]);
            
        end
              
    end
end

File: common/TileAnalysis.py
import datetime as datetime
import numpy as np
from Utils import Str2Date
from PortfolioProviders import PortfolioProviders
from ReturnSeries import ReturnSeries
from Factor import Factor
import Factorlib.BPCalc
import Stock
from QTimeSeries import QTimeSeries
import PortfolioProvider


class TileAnalysis:
    """ The tile analysis is to put all stocks into equally sized groups and the number of groups is defined by Tile
        and then construct tile portfolios for each tile either equally or cap-weighted.
        analyze the time series returns of each tile portfolio
    """

    def __init__(self, dates, univPP, numTiles=5):
        if len(dates) and type(dates[0]) is str:
            self.__dates = [Str2Date(d) for d in dates]
        self.__univPP = univPP
        self.__numTiles = numTiles

    def run(self, factor=None, bmDemean=False):
        # Initialize ReturnSeries
        periods = len(self.__dates) - 1
        topTileRets = ReturnSeries([], [])
        botTileRets = ReturnSeries([], [])
        spreads = ReturnSeries([], [])

        # Force factor to exist for bmDemean to be True
        if(factor and (bmDemean == True)):
            bmDemean = True

        for i in range(periods):
            # Get portfolio and it's stockIDs
            dt = self.__dates[i]
            nextDt = self.__dates[i+1]
            portfolio = self.__univPP.GetPortfolioAsofFixed(dt, 0)
            stockIDs = portfolio.GetStockIDs()

            # Calculate Factor scores and sort them
            scores = np.zeros(len(stockIDs))
            for j in range(len(stockIDs)):
                scores[j] = factor.GetScore(stockIDs, dt, True)

            # Find fraction of scores that are np.nan
            numNans = np.sum(np.isnan(scores))
            fracNans = numNans / float(len(scores))
            try:
                if fracNans > 0.5:
                    raise ValueError('Half of stocks have NaN scores')
            except ValueError as e:
                print 'Error: ', e

            scoreSortIdx = np.argsort(scores)  # sort ascend or descend ??; is NaN on the top or down the bottom?

            # Initialize lists of stock returns
            numStocksInTile = np.floor(len(scores) / float(self.__numTiles))
            topTileStkRets = np.zeros(numStocksInTile)
            botTileStkRets = np.zeros(numStocksInTile)

            for j in range(numStocksInTile):
                topTileStkRets[j] = Stock.ByWindID(stockIDs[scoreSortIdx[j+numNans]]).TotalReturnInRange_Bk(dt, nextDt)
                botTileStkRets[j] = Stock.ByWindID(stockIDs[scoreSortIdx[numStocksInTile-j+1]]).TotalReturnInRange_Bk(dt, nextDt)

            # equal weighted for now, later need to check whether it is cap weighted or equal weighted
            topRet = np.nanmean(topTileStkRets)
            botRet = np.nanmean(botTileStkRets)

            if (bmDemean):
                bmRet = self.__univPP.TotalReturnInRange_Bk(dt, nextDt)
                topRet = topRet - bmRet
                botRet = botRet - bmRet

            topTileRets.Add(nextDt, topRet)
            botTileRets.Add(nextDt, botRet)
            spreads.Add(nextDt, topRet - botRet)

            print(['Top AnnMean, ', str(topTileRets.AnnMean)])
            print(['Top AnnStd,  ', str(topTileRets.AnnStd)])
            print(['Top SR,      ', str(topTileRets.SR)])

            print(['Bot AnnMean, ', str(botTileRets.AnnMean)])
            print(['Bot AnnStd,  ', str(botTileRets.AnnStd)])
            print(['Bot SR,      ', str(botTileRets.SR)])

            print(['Spread AnnMean, ', str(spreads.AnnMean)])
            print(['Spread AnnStd,  ', str(spreads.AnnStd)])
            print(['Spread SR,      ', str(spreads.SR)])




if __name__ == '__main__':
    tileAnalysis = TileAnalysis(['20130101', '20140101', '20150101'], PortfolioProviders.getA50(), 5)
    BP = Factor('BP', 'Book/Price', BPCalc, zz800PP)
    tileAnalysis.run(BP, True)
